File: Languages/Php/Portuguese/Logic/Logic.php
<?php

// 1. Variáveis e Tipos de Dados
// Variáveis são como "caixas" onde armazenamos informações. Cada caixa tem um tipo específico.

$idade = 25;            // Tipo inteiro: armazena números inteiros
$altura = 1.75;         // Tipo ponto flutuante: armazena números com casas decimais
$nome = "João";         // Tipo texto (String): armazena uma sequência de caracteres
$estudante = true;      // Tipo lógico (booleano): armazena verdadeiro ou falso

echo "Idade: " . $idade . "\n";
echo "Altura: " . $altura . "\n";
echo "Nome: " . $nome . "\n";
echo "Estudante: " . ($estudante ? "Sim" : "Não") . "\n";

// 2. Operadores
// Operadores aritméticos, de comparação e lógicos.

// Operadores Aritméticos
$soma = 10 + 5;  // Adição
$resto = 10 % 3; // Módulo (resto da divisão)
echo "Soma: " . $soma . "\n"; 
echo "Resto: " . $resto . "\n";

// Operadores de Comparação
$resultado = 10 > 5; // Verifica se 10 é maior que 5
$iguais = 5 == 5;    // Verifica se 5 é igual a 5
echo "10 > 5: " . ($resultado ? "Verdadeiro" : "Falso") . "\n";
echo "5 == 5: " . ($iguais ? "Verdadeiro" : "Falso") . "\n";

// Operadores Lógicos
$eCondicao = (10 > 5) && (5 < 10);  // 'E' lógico: ambas as condições precisam ser verdadeiras
$ouCondicao = (10 > 5) || (5 > 10); // 'OU' lógico: uma das condições precisa ser verdadeira
echo "Condição E: " . ($eCondicao ? "Verdadeiro" : "Falso") . "\n";
echo "Condição OU: " . ($ouCondicao ? "Verdadeiro" : "Falso") . "\n";

// 3. Estruturas Condicionais
// Condições para tomar decisões no programa.

$idadePessoa = 20;
if ($idadePessoa >= 18) {
    echo "Você é maior de idade.\n";
} else {
    echo "Você é menor de idade.\n";
}

// Switch Case: alternativa ao if-else quando temos várias condições possíveis
$numero = 2;
switch ($numero) {
    case 1:
        echo "É um\n";
        break;
    case 2:
        echo "É dois\n";
        break;
    default:
        echo "Outro número\n";
}

// 4. Laços de Repetição (Loops)
// Usados quando queremos repetir um bloco de código várias vezes.

for ($i = 1; $i <= 5; $i++) { 
    echo $i . "\n";  // Vai imprimir 1, 2, 3, 4, 5
}

$j = 1;
while ($j <= 5) {
    echo $j . "\n";  // Vai imprimir 1, 2, 3, 4, 5
    $j++;  // Incrementa o valor de j
}

$k = 1;
do {
    echo $k . "\n";  // Vai imprimir 1, 2, 3, 4, 5
    $k++;
} while ($k <= 5);

// 5. Funções (Métodos)
// Funções são usadas para organizar e reutilizar blocos de código.

function saudacao() {
    echo "Olá, seja bem-vindo!\n";
}

function soma($a, $b) {
    return $a + $b;  // Retorna a soma dos dois números
}

saudacao();  // Chama a função saudacao
$resultadoSoma = soma(5, 3);  // Chama a função soma
echo "Resultado da soma: " . $resultadoSoma . "\n";

// 6. Entrada e Saída de Dados
// Usado para interagir com o usuário, capturando entradas e exibindo saídas.

function entradaSaida() {
    $nome = readline("Digite seu nome: ");  // Captura o nome do usuário
    echo "Olá, " . $nome . "\n";
}

// 7. Manipulação de Strings
// Strings são sequências de caracteres, úteis para trabalhar com texto.

function manipulacaoStrings() {
    $nome = "João";
    $saudacao = "Olá, " . $nome . "!";  // Concatenando strings
    echo $saudacao . "\n";  // Exibe: Olá, João!

    $palavra = "programação";
    $tamanho = strlen($palavra);  // Verifica o tamanho da string
    $contem = strpos($palavra, "grama") !== false;  // Verifica se a palavra contém "grama"
    echo "Tamanho da palavra: " . $tamanho . "\n";
    echo "Contém 'grama': " . ($contem ? "Sim" : "Não") . "\n";
}

// 8. Arrays
// Arrays são usados para armazenar múltiplos valores do mesmo tipo em uma única variável.

function exemploArrays() {
    $numeros = [1, 2, 3, 4, 5];  // Declaração e inicialização de um array
    echo "Primeiro número: " . $numeros[0] . "\n";  // Acessa o primeiro valor
    echo "Último número: " . $numeros[count($numeros) - 1] . "\n";  // Acessa o último valor
}

// 9. Classe e Objetos
// Criando uma classe e instanciando objetos dela.

class Pessoa {
    public $nome;
    public $idade;

    function __construct($nome, $idade) {
        $this->nome = $nome;
        $this->idade = $idade;
    }

    function apresentar() {
        echo "Olá, meu nome é " . $this->nome . " e tenho " . $this->idade . " anos.\n";
    }
}

function exemploClasseObjetos() {
    $pessoa1 = new Pessoa("Nico", 30);  // Criação de um objeto da classe Pessoa
    $pessoa1->apresentar();  // Chamada do método
}

// 10. Conceito de `this`
// O `this` é usado para referir-se ao objeto atual dentro de um método da classe.

function exemploThis() {
    class Carro {
        public $modelo;

        function __construct($modelo) {
            $this->modelo = $modelo;  // 'this' faz referência ao atributo da classe
        }

        function mostrarModelo() {
            echo "Modelo do carro: " . $this->modelo . "\n";
        }
    }

    $carro = new Carro("Fusca");
    $carro->mostrarModelo();  // Exibe: Modelo do carro: Fusca
}

// 11. Herança
// Criando uma classe filha que herda de uma classe pai.

class Animal {
    public $nome;

    function __construct($nome) {
        $this->nome = $nome;
    }

    function fazerSom() {
        echo "O animal faz um som.\n";
    }
}

class Cachorro extends Animal {
    function fazerSom() {
        echo "O cachorro late.\n";
    }
}

function exemploHeranca() {
    $cachorro = new Cachorro("Rex");
    $cachorro->fazerSom();  // O método sobreposto é chamado
}

// 12. Encapsulamento
// Usando modificadores de acesso para proteger dados.

function exemploEncapsulamento() {
    class ContaBancaria {
        private $saldo;  // Atributo privado

        function __construct() {
            $this->saldo = 0;
        }

        function depositar($valor) {
            if ($valor > 0) {
                $this->saldo += $valor;
            }
        }

        function getSaldo() {
            return $this->saldo;  // Acesso ao saldo através de um método público
        }
    }

    $conta = new ContaBancaria();
    $conta->depositar(500);
    echo "Saldo da conta: " . $conta->getSaldo() . "\n";
}

// 13. Exceções
// Tratamento de erros durante a execução do programa.

function exemploExcecao() {
    try {
        $resultado = 10 / 0;  // Divisão por zero
    } catch (Exception $e) {
        echo "Erro: " . $e->getMessage() . "\n";
    }
}

// 14. Trabalhando com `ArrayList`
// O ArrayList é uma coleção que pode crescer dinamicamente.

function exemploArrayList() {
    $lista = ["Java", "Python", "C++"];

    echo "Tamanho da lista: " . count($lista) . "\n";  // Exibe o tamanho da lista
    echo "Primeiro item: " . $lista[0] . "\n";  // Acessa o primeiro item da lista
}

?>

File: Languages/Swift/Portuguese/Logic/Logic.swift
// 1. Variáveis e Tipos de Dados
let idade = 25            // Tipo inteiro: armazena números inteiros
let altura = 1.75         // Tipo ponto flutuante: armazena números com casas decimais
let nome = "João"         // Tipo texto (String): armazena uma sequência de caracteres
let estudante = true      // Tipo lógico (booleano): armazena verdadeiro ou falso

print("Idade: \(idade)")
print("Altura: \(altura)")
print("Nome: \(nome)")
print("Estudante: \(estudante)")

// 2. Operadores
let soma = 10 + 5  // Adição
let resto = 10 % 3 // Módulo (resto da divisão)
print("Soma: \(soma)")
print("Resto: \(resto)")

// Operadores de Comparação
let resultado = 10 > 5  // Verifica se 10 é maior que 5
let iguais = 5 == 5     // Verifica se 5 é igual a 5
print("10 > 5: \(resultado)")
print("5 == 5: \(iguais)")

// Operadores Lógicos
let eCondicao = (10 > 5) && (5 < 10)  // 'E' lógico: ambas as condições precisam ser verdadeiras
let ouCondicao = (10 > 5) || (5 > 10) // 'OU' lógico: uma das condições precisa ser verdadeira
print("Condição E: \(eCondicao)")
print("Condição OU: \(ouCondicao)")

// 3. Estruturas Condicionais
let idadePessoa = 20
if idadePessoa >= 18 {
    print("Você é maior de idade.")
} else {
    print("Você é menor de idade.")
}

// Switch Case
let numero = 2
switch numero {
case 1:
    print("É um")
case 2:
    print("É dois")
default:
    print("Outro número")
}

// 4. Laços de Repetição (Loops)
for i in 1...5 {
    print(i)  // Vai imprimir 1, 2, 3, 4, 5
}

var j = 1
while j <= 5 {
    print(j)  // Vai imprimir 1, 2, 3, 4, 5
    j += 1    // Incrementa o valor de j
}

var k = 1
repeat {
    print(k)  // Vai imprimir 1, 2, 3, 4, 5
    k += 1
} while k <= 5

// 5. Funções (Métodos)
func saudacao() {
    print("Olá, seja bem-vindo!")
}

func soma(a: Int, b: Int) -> Int {
    return a + b  // Retorna a soma dos dois números
}

saudacao()  // Chama a função saudacao
let resultadoSoma = soma(a: 5, b: 3)  // Chama a função soma
print("Resultado da soma: \(resultadoSoma)")

// 6. Entrada e Saída de Dados
func entradaSaida() {
    print("Digite seu nome: ")
    if let nome = readLine() {
        print("Olá, \(nome)")
    }
}

// 7. Manipulação de Strings
func manipulacaoStrings() {
    let nome = "João"
    let saudacao = "Olá, \(nome)!"  // Concatenando strings
    print(saudacao)  // Exibe: Olá, João!

    let palavra = "programação"
    let tamanho = palavra.count  // Verifica o tamanho da string
    let contem = palavra.contains("grama")  // Verifica se a palavra contém "grama"
    print("Tamanho da palavra: \(tamanho)")
    print("Contém 'grama': \(contem)")
}

// 8. Arrays
func exemploArrays() {
    let numeros = [1, 2, 3, 4, 5]  // Declaração e inicialização de um array
    print("Primeiro número: \(numeros[0])")  // Acessa o primeiro valor
    print("Último número: \(numeros[numeros.count - 1])")  // Acessa o último valor
}

// 9. Classe e Objetos
class Pessoa {
    var nome: String
    var idade: Int

    init(nome: String, idade: Int) {
        self.nome = nome
        self.idade = idade
    }

    func apresentar() {
        print("Olá, meu nome é \(self.nome) e tenho \(self.idade) anos.")
    }
}

func exemploClasseObjetos() {
    let pessoa1 = Pessoa(nome: "Nico", idade: 30)  // Criação de um objeto da classe Pessoa
    pessoa1.apresentar()  // Chamada do método
}

// 10. Conceito de `this`
func exemploThis() {
    class Carro {
        var modelo: String

        init(modelo: String) {
            self.modelo = modelo  // 'self' faz referência ao atributo da classe
        }

        func mostrarModelo() {
            print("Modelo do carro: \(self.modelo)")
        }
    }

    let carro = Carro(modelo: "Fusca")
    carro.mostrarModelo()  // Exibe: Modelo do carro: Fusca
}

// 11. Herança
class Animal {
    var nome: String

    init(nome: String) {
        self.nome = nome
    }

    func fazerSom() {
        print("O animal faz um som.")
    }
}

class Cachorro: Animal {
    override func fazerSom() {
        print("O cachorro late.")
    }
}

func exemploHeranca() {
    let cachorro = Cachorro(nome: "Rex")
    cachorro.fazerSom()  // O método sobreposto é chamado
}

// 12. Encapsulamento
func exemploEncapsulamento() {
    class ContaBancaria {
        private var saldo: Double  // Atributo privado

        init() {
            self.saldo = 0
        }

        func depositar(valor: Double) {
            if valor > 0 {
                self.saldo += valor
            }
        }

        func getSaldo() -> Double {
            return self.saldo  // Acesso ao saldo através de um método público
        }
    }

    let conta = ContaBancaria()
    conta.depositar(valor: 500)
    print("Saldo da conta: \(conta.getSaldo())")
}

// 13. Exceções
func exemploExcecao() {
    do {
        let resultado = 10 / 0  // Divisão por zero
    } catch {
        print("Erro: \(error)")
    }
}

// 14. Trabalhando com Arrays
func exemploArrayList() {
    var lista = ["Java", "Python", "C++"]

    print("Tamanho da lista: \(lista.count)")  // Exibe o tamanho da lista
    print("Primeiro item: \(lista[0])")  // Acessa o primeiro item da lista
}
